services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java

```diff
@@ -261,54 +263,71 @@ public static class CRLFileLoader extends CRLLoaderImpl {
         private final KeycloakSession session;
         private final String cRLPath;
         private final LdapContext ldapContext;
+        private final boolean abortIfNonUpdated;
 
-        public CRLFileLoader(KeycloakSession session, String cRLPath) {
-            this.session = session;
-            this.cRLPath = cRLPath;
-            ldapContext = new LdapContext();
+        public CRLFileLoader(KeycloakSession session, String cRLPath, boolean abortIfNonUpdated) {
+            this(session, cRLPath, abortIfNonUpdated, new LdapContext());
         }
 
-        public CRLFileLoader(KeycloakSession session, String cRLPath, LdapContext ldapContext) {
+        public CRLFileLoader(KeycloakSession session, String cRLPath, boolean abortIfNonUpdated, LdapContext ldapContext) {
             this.session = session;
             this.cRLPath = cRLPath;
+            this.abortIfNonUpdated = abortIfNonUpdated;
             this.ldapContext = ldapContext;
 
             if (ldapContext == null)
                 throw new NullPointerException("Context cannot be null");
         }
 
         public Collection<X509CRL> getX509CRLs() throws GeneralSecurityException {
-            CertificateFactory cf = CertificateFactory.getInstance("X.509");
-            Collection<X509CRL> crlColl = null;
+            CrlStorageProvider crlCache = session.getProvider(CrlStorageProvider.class);
+            X509CRL crl = null;
+            if (cRLPath != null) {
+                crl = crlCache.get(cRLPath, this::loadCRL);
+            }
+
+            if (crl == null) {
+                throw new GeneralSecurityException(String.format("Unable to load CRL from \"%s\"", cRLPath));
+            }
+
+            if (crl.getNextUpdate() != null && crl.getNextUpdate().compareTo(new Date(Time.currentTimeMillis())) < 0) {
+                final String message = String.format("CRL from '%s' is not refreshed. Next update is %s.", cRLPath, crl.getNextUpdate());
+                logger.warn(message);
+                if (abortIfNonUpdated) {
+                    throw new GeneralSecurityException(message);
+                }
+            }
 
+            return Collections.singletonList(crl);
+        }
+
+        private X509CRL loadCRL() throws GeneralSecurityException {
+            CertificateFactory cf = CertificateFactory.getInstance("X.509");
```

Comment:
check if `cRLPath` is null first to avoid lookups.

Reply:
I just removed the if here as it was done before in the previous method.